lambda/dissect_letrec.ml

```diff
@@ -0,0 +1,592 @@
+  (* Prefix of the expression that does not depends on any recursive part.
+     This is presented as a function for easy 'concatenation':
+     to append 'expr': [fun ~tail -> Lsequence (expr, letrec.pre ~tail)] *)
+  effects : Lambda.lambda;
+  (* Effects that are applied afterwards. *)
+  functions : (Ident.t * Lambda.lfunction) list;
+  substitution : Ident.t Ident.Map.t;
+  (* Alias to recursive variables should be forbidden, to prevent
+     non-productive definition like 'let rec a = a'. But some aliases
+     are allowed, for instance 'let rec a = let c = b in 1 :: b and b = 2 :: a'.
+     The simplest way to handle those aliases is simply
+     to apply a substitute of all these aliases afterward. *)
+  letbound : Ident.Set.t;
+  (* Set of known immutable variables. Mutable ones cannot define aliases *)
+}
+
+type let_def = {
+  let_kind : Lambda.let_kind;
+  value_kind : Lambda.value_kind;
+  ident : Ident.t;
+}
+
+exception Bug
+
+let () = ignore Bug
+
+let lsequence (lam1, lam2) =
+  match lam1 with
+  | Lsequence (lam, Lconst (Const_pointer 0)) -> Lsequence (lam, lam2)
+  | Lconst (Const_pointer 0) -> lam2
+  | _ -> Lsequence (lam1, lam2)
+
+let caml_update_dummy_prim =
+  Primitive.simple ~name:"caml_update_dummy" ~arity:2 ~alloc:true
+
+let update_dummy var expr =
+  Lprim (Pccall caml_update_dummy_prim, [Lvar var; expr], Location.none)
+
+let build_block let_def size block_type expr letrec =
+  { letrec with
+    blocks = (let_def.ident, { block_type; size }) :: letrec.blocks;
+    effects = Lsequence (update_dummy let_def.ident expr, letrec.effects);
+  }
+
+let is_simple (lam:Lambda.lambda) =
+  match lam with
+  | Lvar _
+  | Lconst _ -> true
+  | _ -> false
+
+let dead_code lam letrec =
+  (* Some cases generate code without effects, and bound to nothing. We use this
+     function to insert it as [Lsequence] in [effects], for documentation.
+     It would be correct to discard and just return [letrec] instead. *)
+  { letrec with effects = lsequence (lam, letrec.effects) }
+
+(* We desconstruct the let-rec into a description *)
+
+let rec prepare_letrec
+    (recursive_set:Ident.Set.t)
+    (* Variables that depends on the let-rec bound variables *)
+    (current_let:let_def option)
+    (* The variable to which the current expression is bound.
+       current_let.ident is part of recursive_set *)
+    (lam:Lambda.lambda)
+    (letrec:letrec) =
+  match lam with
+  | Lfunction funct -> begin
+      match current_let with
+      | Some current_let when Ident.Set.mem current_let.ident recursive_set ->
+          { letrec with functions =
+                          (current_let.ident, funct) :: letrec.functions }
+      | Some current_let ->
+          (* If the currently bound function does not depend on any
+             recursive variable *)
+          let pre ~tail : Lambda.lambda =
+            Llet (current_let.let_kind, current_let.value_kind,
+                  current_let.ident, lam, letrec.pre ~tail)
```

Comment:
Note: with this definition, the call that runs last performs its effects first. This explain the slightly-odd implementation of the `Lsequence(lam1, lam2)` for example, that looks essentially like

```ocaml
let letrec = dissect lam2 in
let letrec = dissect lam1 in
letrec
```
(state-passing-style, in reverse order)

If we had written instead

```ocaml
let pre ~tail =
  letrec.pre (Llet (..., ~tail))
```

then we could write the calls in the order of evaluation, instead of the reverse order.

(It might be nice to try this to see if it's a big change and/or if it really makes the code nicer.)